asm-to-pil/src/vm_to_constrained.rs

```diff
@@ -454,6 +456,86 @@ impl<T: FieldElement> ASMPILConverter<T> {
         res
     }
 
+    fn handle_instr_declaration_params(&mut self, m: &mut Machine<T>) {
+        for l in &mut m.links {
+            let lhs = &l.params;
+            let rhs = &mut l.to.params;
+            // lhs params must all be assignment registers when mapping instruction to operation
+            lhs.inputs.iter().chain(lhs.outputs.iter()).for_each(|p| {
+                assert!(p.ty.is_none());
+                assert!(self
+                    .registers
+                    .get(&p.name)
+                    .is_some_and(|r| r.ty.is_assignment()));
+            });
+
+            // if rhs is not empty, check it's valid
+            if !rhs.is_empty() {
+                // rhs params must either be assignment registers declared in the lhs or write registers
+                rhs
+                .inputs
+                .iter_mut()
+                .chain(rhs.outputs.iter_mut())
+                .for_each(|p| {
+                    let reg = self.registers.get(&p.name).expect("rhs params must be registers");
+                    assert!(p.ty.is_none());
+                    match reg.ty {
+                        RegisterTy::Assignment => {
+                            let mut lhs_params = lhs.inputs.iter().chain(lhs.outputs.iter());
```

Comment:
Maybe it would be useful to have a function that returns an iterator over inputs and outputs?

Reply:
Good idea